from typing import Callable
from functools import wraps


def exception_handler(handler: Callable[[str], None], process_name: str = ''):
    def decorate(fn):
        @wraps(fn)
        def wrapper(*args, **kwargs):
            try:
                result = fn(*args, **kwargs)
            except Exception as e:
                handler(f'Erro ao executar processo {process_name}:\n{str(e)}')
                raise e
            return result
        return wrapper
    return decorate
